File: 1.ml
let rec solve curr = match curr with
    | 0 -> 0
    | _ when (curr mod 3 = 0) || (curr mod 5 = 0) -> curr + (solve (curr - 1))
    | _ -> solve (curr - 1);;

let rec solve2 curr = match curr with
    | 0 -> 0
    | _ -> (if (curr mod 3 = 0) || (curr mod 5 = 0) then curr else 0) +
    solve2 (curr - 1);;

let rec range a b = match a with
    | _ when a > b -> []
    | _ -> a :: (range (a + 1) b);;

let solve3 curr =
    List.fold_left (fun acc x -> acc + x) 0
    (List.filter (fun x -> x mod 3 = 0 || x mod 5 = 0) (range 1 curr));;

Printf.printf "%d\n" (solve 999);
Printf.printf "%d\n" (solve2 999);
Printf.printf "%d\n" (solve3 999)

File: 1.clj
(defn solve [curr]
  (cond
    (zero? curr) 0
    (or (zero? (rem curr 3)) (zero? (rem curr 5)))
    (+ curr (solve (dec curr)))
    :else (solve (dec curr))))

(defn solve2 [curr]
  (cond
    (zero? curr) 0
    :else (+ (if (or (zero? (rem curr 3)) (zero? (rem curr 5)))
               curr
               0)
             (solve2 (dec curr)))))

(defn solve3 [curr]
  (reduce + (filter #(or (zero? (rem % 3)) (zero? (rem % 5)))
                    (range (inc curr)))))

(println (solve 999))
(println (solve2 999))
(println (solve3 999))
